onnxruntime/core/providers/dml/DmlExecutionProvider/src/TensorDesc.cpp

```diff
@@ -315,3 +315,60 @@ void TensorDesc::SetDimensionCount(uint32_t newDimensionCount, TensorAxis alignm
     }
     m_bufferTensorDesc.DimensionCount = newDimensionCount;
 }
+
+// Uses dimensionMapping to reorder m_sizes and m_strides to match specific Tensor layout
+void TensorDesc::PermuteDimensions(gsl::span<const uint32_t> dimensionMapping, const TensorAxis alignment)
+{
+    EnsureMinimumDimensionCount(static_cast<uint32_t>(dimensionMapping.size()), alignment);
+    EnsureStridesExist(static_cast<uint32_t>(dimensionMapping.size()), alignment);
+    PermuteArray(dimensionMapping, alignment);
+}
+
+// Shuffle m_sizes and m_strides according to the indexes pointed by dimensionMapping
+void TensorDesc::PermuteArray(gsl::span<const uint32_t> dimensionMapping, TensorAxis alignment)
+{
+    std::vector<uint32_t> tempSizes(dimensionMapping.size());
+    std::vector<uint32_t> tempStrides(dimensionMapping.size());
+    // Right alignment values are shifted to the end
+    int alignmentOffset = (alignment == TensorAxis::RightAligned) ? MaximumDimensionCount - static_cast<int>(dimensionMapping.size()) : 0;
+
+    for (size_t i = 0; i < dimensionMapping.size(); i++)
+    {
+        tempSizes[i] = m_sizes[dimensionMapping[i] + alignmentOffset];
+        tempStrides[i] = m_strides[dimensionMapping[i] + alignmentOffset];
+    }
+
+    std::copy(tempSizes.begin(), tempSizes.end(), m_sizes + alignmentOffset);
+    std::copy(tempStrides.begin(), tempStrides.end(), m_strides + alignmentOffset);
+    m_bufferTensorDesc.Sizes = m_sizes;
+    m_bufferTensorDesc.Strides = m_strides;
+}
+
+void TensorDesc::EnsureStridesExist(uint32_t dimensionCount, TensorAxis alignment)
```

Comment:
>TensorAxis alignment [](http://example.com/codeflow?start=61&length=20)

The alignment parameter is not needed for `EnsureStridesExist`, because the size has already been set by `EnsureMinimumDimensionCount`.